internal/app/apid/main.go

```diff
@@ -0,0 +1,127 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package main
+
+import (
+	"flag"
+	"log"
+	stdlibnet "net"
+	"os"
+	"strings"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/talos-systems/talos/api/api"
+	"github.com/talos-systems/talos/pkg/config"
+	"github.com/talos-systems/talos/pkg/constants"
+	"github.com/talos-systems/talos/pkg/grpc/factory"
+	"github.com/talos-systems/talos/pkg/grpc/tls"
+	"github.com/talos-systems/talos/pkg/net"
+	"github.com/talos-systems/talos/pkg/startup"
+)
+
+var (
+	configPath *string
+	endpoints  *string
+)
+
+func init() {
+	log.SetFlags(log.Lshortfile | log.Ldate | log.Lmicroseconds | log.Ltime)
+
+	configPath = flag.String("config", "", "the path to the config")
+	endpoints = flag.String("endpoints", "", "the IPs of the control plane nodes")
+
+	flag.Parse()
+}
+
+func main() {
+	if err := startup.RandSeed(); err != nil {
+		log.Fatalf("failed to seed RNG: %v", err)
+	}
+
+	provider, err := createProvider()
+	if err != nil {
+		log.Fatalf("failed to create remote certificate provider: %+v", err)
+	}
+
+	ca, err := provider.GetCA()
+	if err != nil {
+		log.Fatalf("failed to get root CA: %+v", err)
+	}
+
+	tlsConfig, err := tls.New(
+		tls.WithClientAuthType(tls.Mutual),
+		tls.WithCACertPEM(ca),
+		tls.WithCertificateProvider(provider),
+	)
+	if err != nil {
+		log.Fatalf("failed to create OS-level TLS configuration: %v", err)
+	}
+
+	machineClient, err := api.NewLocalMachineClient()
+	if err != nil {
+		log.Fatalf("init client: %v", err)
+	}
+
+	osClient, err := api.NewLocalOSClient()
+	if err != nil {
+		log.Fatalf("networkd client: %v", err)
+	}
+
+	interceptorProvider, err := createProvider()
+	if err != nil {
+		log.Fatalf("failed to create remote certificate provider for interceptor: %+v", err)
+	}
+
+	protoProxy := api.NewApiProxy(interceptorProvider)
+
+	err = factory.ListenAndServe(
+		&api.Registrator{
+			MachineClient: machineClient,
+			OSClient:      osClient,
+		},
+		factory.Port(constants.OsdPort),
```

Comment:
I'm a bit confused about this. Is the idea that apid will run on every node?

Reply:
Yeah that's always been the plan -- to replace the frontend with this. I think we've talked about it a bit before and i don't think it's unreasonable to add in an option to control the proxy functionality ( we _should_ just be able to drop the unary interceptor portion and be back to normal ), but honestly I'm not sure I see any benefit to disabling that.

Reply:
I have concerns about the long term effects of this, particularly when we get to implementing RBAC. I think we should run APId on master nodes only since there are already sensitive files on those machines and it’s best practice to not allow workloads on those machines. I’d like to see worker nodes have zero ability to reach out to any other node and use the API.

Reply:
Let's chat about this at the community meeting today. I think this is worth a deeper discussion.